src/istio/mixerclient/check_context.h

```diff
@@ -0,0 +1,197 @@
+                         const istio::mixer::v1::CheckResponse& response) {
+    policy_cache_result_.SetResponse(status, *shared_attributes_->attributes(),
+                                     response);
+  }
+
+  //
+  // Quota Cache Checks
+  //
+
+  bool quotaCheckRequired() const { return !quota_requirements_.empty(); }
+  bool remoteQuotaRequestRequired() const {
+    return remote_quota_check_required_;
+  }
+
+  void checkQuotaCache(QuotaCache& quotaCache) {
+    if (!quotaCheckRequired()) {
+      return;
+    }
+
+    //
+    // Quota is removed from the quota cache iff there is a policy cache hit. If
+    // there is a policy cache miss, then a request has to be sent upstream
+    // anyways, so the quota will be decremented on the upstream response.
+    //
+    quotaCache.Check(*shared_attributes_->attributes(), quota_requirements_,
+                     policyCacheHit(), &quota_cache_result_);
+
+    remote_quota_check_required_ =
+        quota_cache_result_.BuildRequest(allocRequestOnce());
+
+    quota_cache_hit_ = quota_cache_result_.IsCacheHit();
+  }
+
+  void updateQuotaCache(const google::protobuf::util::Status& status,
+                        const istio::mixer::v1::CheckResponse& response) {
+    quota_cache_result_.SetResponse(status, *shared_attributes_->attributes(),
+                                    response);
+  }
+
+  bool quotaCacheHit() const { return quota_cache_hit_; }
+  const google::protobuf::util::Status& quotaStatus() const {
+    return quota_cache_result_.status();
+  }
+
+  //
+  // Upstream request and response
+  //
+
+  void compressRequest(const AttributeCompressor& compressor,
+                       const std::string& deduplication_id) {
+    compressor.Compress(*shared_attributes_->attributes(),
+                        allocRequestOnce()->mutable_attributes());
+    request_->set_global_word_count(compressor.global_word_count());
+    request_->set_deduplication_id(deduplication_id);
+  }
+
+  bool failOpen() const { return fail_open_; }
+
+  const istio::mixer::v1::CheckRequest& request() { return *request_; }
+
+  istio::mixer::v1::CheckResponse* response() {
+    if (!response_) {
+      response_ = google::protobuf::Arena::CreateMessage<
+          istio::mixer::v1::CheckResponse>(&shared_attributes_->arena());
+    }
+    return response_;
+  }
+
+  void setFinalStatus(const google::protobuf::util::Status& status,
+                      bool add_report_attributes = true) {
+    if (add_report_attributes) {
+      utils::AttributesBuilder builder(shared_attributes_->attributes());
+      builder.AddBool(utils::AttributeName::kCheckCacheHit, policy_cache_hit_);
+      builder.AddBool(utils::AttributeName::kQuotaCacheHit, quota_cache_hit_);
+    }
+
+    final_status_ = status;
+  }
+
+  //
+  // CheckResponseInfo (exposed to the top-level filter)
+  //
+
+  virtual const google::protobuf::util::Status& status() const override {
```

Comment:
Nit: this doesn't need to be `virtual` anymore.

Reply:
for my own education:  because of the override keyword, or because it inherits from a class that already declared this virtual?